go-71 context menus

mickedefault opened the pull request:
adds a bunch of context menus

- [x] stage files context menu
- [x] git history details context menu
- [x] stage files text context menu
- [x] diff file context menu filename should copy only filename, not relative path

mickedefault committed:
refactor(theme): extract treeviewitemstyle to separate style

this will allow restyling and applying 'basedon' on the defaulttreeviewitemstyle

mickedefault committed:
refactor(diff): injects parent folder via lookup

ls-tree does not write the relative path, even with --full-name unless the -r[ecursive] command is given. this is because a git tree is a flat list of files without knowledge of its relative position

mickedefault committed:
refactor(commands): registers new global command handlers via globalcommandhandler

mickedefault committed:
refactor(material): moves default treeviewitem template to separate style, allowing basedon logic

mickedefault committed:
feature(diff): adds context menu to diff files

mickedefault committed:
refactor(commands): moves command handlers to application

mickedefault committed:
enhancement(gitfiletree): open directory via contextmenu opens folder correctly

mickedefault committed:
enhancement(stage): add context menu for index file list

mickedefault committed:
enhancement(stage): add context menu for workspace file list

mickedefault committed:
enhancement(material): improve context menu look and feel

calahatras commented on GitOut/Features/Wpf/Commands/Application.cs:
```diff
@@ -0,0 +1,73 @@
+            RevealInExplorer = new CompositeCommand<string>(OnRevealInExplorer);
+        }
+
+        public static CompositeCommand<string>? Open { get; private set; }
```
Don't need setters if the property is only set in ctor.

mickedefault replied:
it's static, so it does :/
```
1>.\GitOut\Features\Wpf\Commands\Application.cs(17,13,17,17):
 error CS0200: Property or indexer 'Application.Open' cannot be assigned to -- it is read only
```

calahatras replied:
Oh right, didn't notice the static.
I always felt it's not the "right way" to assign a static property from a ctor, but I suppose in this case since Application is more or less a singleton it makes it more OK. Is it possible to add a static ctor to this class instead and assign the properties there?

-----

No, because the callbacks use instance data. I'll resolve this

calahatras commented on GitOut/Features/Collections/SortedLazyAsyncCollection.cs:
```diff
@@ -1,29 +1,30 @@
 
         public SortedLazyAsyncCollection(
-            Func<IAsyncEnumerable<T>> factory,
+            Func<RelativeDirectoryPath, IAsyncEnumerable<T>> factory,
```
Maybe we can make this a generic type as well, like a `TContext`, instead of binding it specifically to `RelativeDirectoryPath`.
That way we can keep this collection free from using other features and allow it to be reused. Not sure what we should use if there is no context, if the need arises I guess we can add an overloaded method or class for that .

mickedefault committed:
enhancement(stage): adds icons to context menu items

calahatras commented on GitOut/Features/Wpf/Commands/Application.cs:
```diff
@@ -0,0 +1,73 @@
+
+        private void OnOpen(string? path)
+        {
+            if (path is not null)
```
If we don't want null, can we instead not take a nullable string as parameter? Or is that because of some interface or something? :thinking:

mickedefault replied:
It's compositecommand that enforces it via the ExecuteAsync method 😒 
I wonder if CC can be tweaked though 🤔 

and thus, it was, that it was tweakable #thereifixedit

calahatras commented on GitOut/Features/Git/Log/GitFileTree.xaml:
```diff
@@ -7,14 +7,91 @@
+                                <MenuItem
+                                    Header="Copy filename"
+                                    Command="{x:Static app:Application.Copy}"
+                                    CommandParameter="{Binding FileName, Converter={StaticResource ToStringConverter}}"
```
FileName has a property for path right? No need for the string converter in that case.

mickedefault replied:
it's CompositeCommand that cannot interpret it
![image](https://user-images.githubusercontent.com/38684932/136459113-39a02318-22be-4bb9-ac27-8abbbc70b745.png)

mickedefault replied:
tried to fix it with an implicit string converter, but it's adamant not to convert it 😄

mickedefault committed:
refactor(log): use RootPath instead of WorkingDirectory

mickedefault committed:
fix(contextmenus): consistent use of directory path delimiter

mickedefault committed:
fix(command): convert filename to string

the compositecommand won't automatically convert
FileName to string. need to do it explicitly.

mickedefault committed:
enhancement(log): adds contextmenu to git history details

mickedefault committed:
refactor(lazyasyncenumerable): use TArg instead of concrete class

mickedefault committed:
enhancement(log): adds 'copy relative path' context menu option

mickedefault committed:
fix(compositecommand): don't enforce nullability on compositecommand

calahatras commented on GitOut/App.xaml.cs:
```diff
@@ -41,6 +41,7 @@ public App()
 
             ILogger<App> logger = host.Services.GetRequiredService<ILogger<App>>();
             RegisterExceptionHandlers(logger);
+            _ = host.Services.GetService<Features.Wpf.Commands.Application>();
```
What is this line doing? `GetService` shouldn't have any side-effects, unless this is somehow creating the Application object?

mickedefault replied:
it is creating the application object :)

calahatras commented on GitOut/Features/Git/Files/GitDirectoryViewModel.cs:
```diff
@@ -12,24 +12,31 @@ namespace GitOut.Features.Git.Files
         private bool isExpanded;
 
-        private GitDirectoryViewModel(FileName fileName, RelativeDirectoryPath parent, SortedObservableCollection<IGitFileEntryViewModel> entries)
+        private GitDirectoryViewModel(IGitRepository repository, FileName fileName, RelativeDirectoryPath parent, SortedObservableCollection<IGitFileEntryViewModel> entries)
```
Time for line breaks ^^

calahatras commented on GitOut/Features/Git/Log/LogEntriesViewModel.cs:
```diff
@@ -37,12 +37,23 @@ public LogEntriesViewModel(GitTreeEvent root, IGitRepository repository)
-            _ = logFiles.MaterializeAsync().AsTask();
+            allFiles = new SortedLazyAsyncCollection<IGitFileEntryViewModel, RelativeDirectoryPath>(_ => ListAllFilesAsync(), IGitDirectoryEntryViewModel.CompareItems);
+            var logFiles = new SortedLazyAsyncCollection<IGitFileEntryViewModel, RelativeDirectoryPath>(relativePath =>
+                GitFileEntryViewModelFactory.ListIdAsync(root.Event.Id, repository, relativePath),
```
I think the lambda should be on one line here (move `relativePath` down in other words) to make it more clear that the ctor takes two parameters, one is the lambda and the other is CompareItems.
_Same below btw_

calahatras replied:
Uhm, you only updated the other one ^^

mickedefault committed:
refactor(directoryviewmodel): add line breaks for long lines

mickedefault committed:
refactor(logentriesviewmodel): add line breaks for parameters

mickedefault committed:
fix(stage): add relativpath in statuschangeviewmodel

the relativepath was missing, resulting in directory not being copyable via contextmenu for stage files

calahatras commented on GitOut/Features/Git/Log/GitFileTree.xaml:
```diff
@@ -7,14 +7,91 @@
+                                </MenuItem>
+                                <MenuItem
+                                    Header="Explore file folder"
+                                    Command="{x:Static app:Application.RevealInExplorer}"
```
This command doesn't work for me, it always opens my user folder
The path of RelativePath recently changed I believe, but the full path is needed for the explorer command

calahatras commented on GitOut/Features/Git/Log/GitFileTree.xaml:
```diff
@@ -7,14 +7,91 @@
+                                <MenuItem
+                                    Header="Copy relative path"
+                                    Command="{x:Static app:Application.Copy}"
+                                    CommandParameter="{Binding RelativePath, Converter={StaticResource ToStringConverter}}" >
```
Le space

mickedefault committed:
bug(log): corrects 'explore file folder' parameter

mickedefault committed:
fix(log): formatting

calahatras commented on GitOut/Features/Git/Log/GitFileTree.xaml:
```diff
@@ -7,14 +7,91 @@
+                                <MenuItem
+                                    Header="Copy full path"
+                                    Command="{x:Static app:Application.Copy}"
+                                    CommandParameter="{Binding FullPath}" >
```
Spääce

calahatras commented on GitOut/Features/Git/Log/GitFileTree.xaml:
```diff
@@ -7,14 +7,91 @@
+                                    Header="Explore file folder"
+                                    Command="{x:Static app:Application.RevealInExplorer}"
+                                    CommandParameter="{Binding RelativeDirectory}"
+                                    >
```
Indänt

mickedefault committed:
last fix

mickedefault committed:
last last fix
